src/main/java/org/junitpioneer/jupiter/CartesianEnumSource.java

```diff
@@ -0,0 +1,183 @@
+ */
+@Target({ ElementType.ANNOTATION_TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Repeatable(CartesianEnumSource.CartesianEnumSources.class)
+@ArgumentsSource(CartesianEnumArgumentsProvider.class)
+public @interface CartesianEnumSource {
+
+	/**
+	 * The enum type that serves as the source of the enum constants.
+	 *
+	 * @see #names
+	 * @see #mode
+	 */
+	Class<? extends Enum<?>> value();
+
+	/**
+	 * The names of enum constants to provide, or regular expressions to select
+	 * the names of enum constants to provide.
+	 *
+	 * <p>If no names or regular expressions are specified, all enum constants
+	 * declared in the specified {@linkplain #value enum type} will be provided.
+	 *
+	 * <p>The {@link #mode} determines how the names are interpreted.
+	 *
+	 * @see #value
+	 * @see #mode
+	 */
+	String[] names() default {};
+
+	/**
+	 * The enum constant selection mode.
+	 *
+	 * <p>Defaults to {@link CartesianEnumSource.Mode#INCLUDE INCLUDE}.
+	 *
+	 * @see CartesianEnumSource.Mode#INCLUDE
+	 * @see CartesianEnumSource.Mode#EXCLUDE
+	 * @see CartesianEnumSource.Mode#MATCH_ALL
+	 * @see CartesianEnumSource.Mode#MATCH_ANY
+	 * @see #names
+	 */
+	CartesianEnumSource.Mode mode() default CartesianEnumSource.Mode.INCLUDE;
+
+	/**
+	 * Enumeration of modes for selecting enum constants by name.
+	 */
+	enum Mode {
+
+		/**
+		 * Select only those enum constants whose names are supplied via the
+		 * {@link CartesianEnumSource#names} attribute.
+		 */
+		INCLUDE(CartesianEnumSource.Mode::validateNames, (name, names) -> names.contains(name)),
+
+		/**
+		 * Select all declared enum constants except those supplied via the
+		 * {@link CartesianEnumSource#names} attribute.
+		 */
+		EXCLUDE(CartesianEnumSource.Mode::validateNames, (name, names) -> !names.contains(name)),
+
+		/**
+		 * Select only those enum constants whose names match all patterns supplied
+		 * via the {@link CartesianEnumSource#names} attribute.
+		 *
+		 * @see java.util.stream.Stream#allMatch(java.util.function.Predicate)
+		 */
+		MATCH_ALL(CartesianEnumSource.Mode::validatePatterns,
+				(name, patterns) -> patterns.stream().allMatch(name::matches)),
+
+		/**
+		 * Select only those enum constants whose names match any pattern supplied
+		 * via the {@link CartesianEnumSource#names} attribute.
+		 *
+		 * @see java.util.stream.Stream#anyMatch(java.util.function.Predicate)
+		 */
+		MATCH_ANY(CartesianEnumSource.Mode::validatePatterns,
+				(name, patterns) -> patterns.stream().anyMatch(name::matches));
+
+		private final BiConsumer<CartesianEnumSource, Set<String>> validator;
+		private final BiPredicate<String, Set<String>> selector;
+
+		Mode(BiConsumer<CartesianEnumSource, Set<String>> validator, BiPredicate<String, Set<String>> selector) {
+			this.validator = validator;
+			this.selector = selector;
+		}
+
+		void validate(CartesianEnumSource enumSource, Set<String> names) {
+			notNull(enumSource, "EnumSource must not be null");
```

Comment:
Probably should have tests for `null` values.

Reply:
> Feel free to use `Objects.requireNonNull(Object, String)` or `Preconditions.notNull(Object, String)`.

Originally I used `org.junit.platform.commons.util.Preconditions.notNull` as it's done in the original `@EnumSource`, but the module-build failed as that package is not exported so I created the custom `notNull(Object, String)`.

I also thought about `Objects.requireNonNull` but I was not sure if the `NullPointerException` would have been a good replacement for `PreconditionViolationException`. It might be the latter is picked up in JUnit internals during the execution, but to be honest I didn't look deeply into that and I tried to maintain the style of sibling features.

However, the `notNull` check is invoked only by `CartesianEnumSource.Mode` methods which have package-private visibility and the input values are generated in the `CartesianEnumArgumentsProvider`, so with the current implementation there is no real chance that they can be `null` and the exception is thrown.

What do you think, shall I use `Objects.requireNonNull` instead?

> Probably should have tests for `null` values.

That's tricky because of the point above, the input values are generated by `CartesianEnumArgumentsProvider#accept` and, as far as I can see, there is no way they can be `null` with any behavioral test case.

However, I can create separate unit tests to cover them. Would you suggest going this way?

Reply:
If it can't be `null` then it does not have to be checked for `null`. Or, alternatively, it does not matter what you use to check for `null`.

You could do:
```java
validator.accept(requireNonNull(enumSource), requireNonNull(names));
```

So compact. 😄

Reply:
Theoretically, it makes sense to do null check because it's a package-private api, potentially callable by any other class in the same package. However, I see it as very unlikely.

I guess it's fine if we drop the `PreconditionViolationException` in favor of `NullPointerException`.

> You could do:
> 
> ```java
> validator.accept(requireNonNull(enumSource), requireNonNull(names));
> ```
> 
> So compact. 😄

And this should also make Sonar happy 😄